lib/rubygems/source.rb

```diff
@@ -67,7 +67,9 @@ def <=>(other)
 
       return -1 if !other.uri
 
-      @uri.to_s <=> other.uri.to_s
+      return 1 unless @uri.to_s == other.uri.to_s
+
+      0
```

Comment:
The idea here is that we still want to treat the sources as equal (e.g. return `0`) when the URLs are the same, but we want the first source to have priority over the second if they differ. This causes the original ordering of the sources to be preserved.